File: 2024 February/Check for Children Sum Property in a Binary Tree/main.py
# Link: https://www.geeksforgeeks.org/problems/children-sum-parent/1


class Node:
    def init(self, val):
        self.data = val
        self.left = None
        self.right = None


class Solution:
    def __init__(self):
        self.flag = True

    def is_sum(self, root):
        if root is None:
            return 0
        if root.left is None and root.right is None:
            return root.data
        sum_ = self.is_sum(root.left) + self.is_sum(root.right)
        if sum_ != root.data:
            self.flag = False
        return root.data

    def is_sum_property(self, root):
        self.is_sum(root)
        return int(self.flag)

File: 2024 February/Check for Children Sum Property in a Binary Tree/main.cpp
// Link: https://www.geeksforgeeks.org/problems/children-sum-parent/1

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int data;
    struct Node *left;
    struct Node *right;

    Node(int x)
    {
        data = x;
        left = right = NULL;
    };
};

class Solution
{
public:
    bool flag = 1;
    int isSum(Node *root)
    {
        if (root == NULL)
            return 0;
        if (root->left == NULL && root->right == NULL)
            return root->data;

        int sum = isSum(root->left) + isSum(root->right);
        if (sum != root->data)
            flag = 0;
        return root->data;
    }

    int isSumProperty(Node *root)
    {
        isSum(root);
        return flag;
    }
};